Improve reheat coil sizing when there is no central heating coil

rongpengzhang opened the pull request:
The current reheat coil sizing approach in EnergyPlus is designed assuming there are central heating coils in the primary air systems. It does not apply for the system without central heating coils. This feature aims to improve it to handle various system configurations. Different sizing parameters will be selected depending on the existance of central heating coil, preheating coil, and outdoor air heat exchangers.

rongpengzhang committed:
Check existence of OA heat exchanger

Codes to check whether there is OA heat exchanger or not

rongpengzhang committed:
Check existence of central heating coils

wfbuhl commented:
Code looks great so far!

rongpengzhang committed:
Address reheat coil sizing for different system configurations

rongpengzhang committed:
Recalculate Humidity Ratio for reheat coil sizing

rongpengzhang committed:
Design Document

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 83811630_ReheatCoilSizingImprove

rongpengzhang committed:
Update ReheatCoilSizingWithNoCentralHC.md

rongpengzhang committed:
Address comments from sizing call

Directly modify the value of
TermUnitFinalZoneSizing().DesHeatCoilInTempTU which is at a root level

rongpengzhang committed:
Unit Test

(1) Develop unit test
(2) Some minor changes to address build warnings

rongpengzhang committed:
Address errors in example file test

There is 1 example file failed in the test: DOASDualDuctSchool.idf

rongpengzhang committed:
Update ReheatCoilSizingWithNoCentralHC.md

EnergyArchmage commented on src/EnergyPlus/SimAirServingZones.cc:
```diff
@@ -6252,6 +6296,151 @@ namespace SimAirServingZones {
+			OutAirFrac = FinalSysSizing( IndexAirLoop ).DesOutAirVolFlow / FinalSysSizing( IndexAirLoop ).DesHeatVolFlow;
+			OutAirFrac = min( 1.0, max( 0.0, OutAirFrac ) );
+
+			ReheatCoilInTempForSizing = OutAirFrac * FinalSysSizing( IndexAirLoop ).PreheatTemp + ( 1 - OutAirFrac ) * FinalSysSizing( IndexAirLoop ).HeatRetTemp;
```
This calculation is only going to be correct if the OA and return air have the same humidity ratio.  I think we usually do these sort of moist mixed air calculations using enthalpy and then convert back to drybulb.

EnergyArchmage commented:
One thing that is missing here is some kind of output to inform the user that something happened.  When this new feature kicks in, values the user entered in the Sizing:System object are discarded and replaced with something else.  This information needs to be reported somehow, but only when the values were changed (so you might want a bool flag to the data structure to record that the override took place).   I think it would be sufficient to send output to the EIO that would come out like this:

```
System Sizing Information, VAV_1, Central Heating Design Supply Air Temperature [C], < value for new dynamically calculated dyrbulb >
System Sizing Information, VAV_1, Central Heating Design Supply Air Humidity Ratio [kgWater/kgDryAir], <value for new dynamically calculated humidity ratio>
```

I would guess that ReportSysSizing routine has already executed. So you could likely reuse the EIO data header already written out there.

mjwitte commented:
This would be useful reporting straight from the sizing routines no matter method is used (central heating, preheat, mixed).  I don't see a need for tracking if this was changed.

rongpengzhang committed:
Address comments from Brent

1. Provide more information to the user on the determination of coil
inlet air temperature/humidity ratio for sizing
(1) add more descriptions of the corresponding fields in the
"Sizing:System" object in IO reference, notifying users that different
values may be used in zone component sizing depending on the exisitance
of central heating coils.
(2) send the following output to the EIO file:
" Component Sizing Information, AirTerminal:SingleDuct:VAV:Reheat,
SPACE4-1 VAV REHEAT, Design Size Reheat Coil Sizing Inlet Air
Temperature [C], 16.70000 "
" Component Sizing Information, AirTerminal:SingleDuct:VAV:Reheat,
SPACE4-1 VAV REHEAT, Design Size Reheat Coil Sizing Inlet Air Humidity
Ratio [kgWater/kgDryAir], 8.00000E-003"
I put these outputs as "Component Sizing Information" instead of "System
Sizing Information", because in this feature we only changed the value
of zone sizing parameter TermUnitFinalZoneSizing( ).DesHeatCoilInTempTU.
There is no change of the system sizing arrays.

2. Update the method to calculate the mixed air temperature, taking into
account the impact of humidity ratio difference. Unit test is updated
accordingly.

rongpengzhang commented:
Hi @EnergyArchmage and @mjwitte,

Thanks for the advice. I have updated the codes to address your comments. More specifically:
1. Provide more information to the user on the determination of coil inlet air temperature/humidity ratio 
   (1) add more descriptions of the corresponding fields in the "Sizing:System" object in IO reference, notifying users that different values may be used in zone component sizing depending on the exisitance of central heating coils.
   (2) send the following output to the EIO file, no matter there is or not central heating coils:
   _" Component Sizing Information, AirTerminal:SingleDuct:VAV:Reheat, SPACE4-1 VAV REHEAT, Design Size Reheat Coil Sizing Inlet Air Temperature [C], 16.70000 "_
   _" Component Sizing Information, AirTerminal:SingleDuct:VAV:Reheat, SPACE4-1 VAV REHEAT, Design Size Reheat Coil Sizing Inlet Air Humidity Ratio [kgWater/kgDryAir], 8.00000E-003"_
   I put these outputs as "Component Sizing Information" instead of "System Sizing Information", because in this feature we only changed the value of zone sizing parameter **TermUnitFinalZoneSizing( ).DesHeatCoilInTempTU**. There is no change of the system sizing arrays. 
2. Update the method to calculate the mixed air temperature, taking into account the impact of humidity ratio difference. Unit test is updated accordingly.

rongpengzhang commented:
The failed example files are due to the change of eio files.

nrel-bot-3 commented:
@rongpengzhang @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rongpengzhang @lgentile it has been 14 days since this pull request was last updated.

rongpengzhang commented:
@Myoldmopar Can you please get this branch merged? Brent finished review and approved this in Nov. 2015. It didn't get merged because of the bug-fix release. Happy New Year!

Myoldmopar committed:
Remove member constructors

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 83811630_ReheatCoilSizingImprove

Myoldmopar commented:
@rongpengzhang Yeah, this seems fine.  I went in and cleaned up a bunch of unused member constructors in DataAirSystems.hh (59d3a03).  Whenever you are making changes to a struct, try to remove the member constructor.  

I then pulled in the latest develop.  CI will run this morning and this should be able to merge in later this afternoon.

Myoldmopar commented:
@mjwitte I took a look over the changes, and don't see any red flags.  I'm relying a bit on the review that occurred in October-ish, though.  The unit test, @rongpengzhang, just yes.  The coverage on the two new functions you added are at [100%](http://energyplus.s3-website-us-east-1.amazonaws.com/coverage/2016-01-05-EnergyPlus-7f1dff9486-x86_64-gcc-4.8UTCD/EnergyPlus/SimAirServingZones.cc.gcov.html).  The diffs all seem as expected.  I'll build and run and spot check some things, but this should be good to go in soon.

mbadams5 commented on tst/EnergyPlus/unit/SimAirServingZones.unit.cc:
```diff
@@ -0,0 +1,170 @@
+#include <DataSystemVariables.hh>
+#include <DataZoneEquipment.hh>
+#include <DesiccantDehumidifiers.hh>
+#include <EMSManager.hh>
```
@rongpengzhang Do you really need ALL these includes? I think you only need a small fraction of C++ Headers, ObjexxFCL Headers, and EnergyPlus Headers. Not only does this help reduce clutter, but also (marginally) helps build times.

rongpengzhang replied:
HI @mbadams5  You are right, my unit tests don't need to include all these heads, which I copied from "SimAirServingZones.cc". The reason I included all of them here is to save my time and to facilitate other developers who may add more unit tests for "SimAirServingZones.cc". "SimAirServingZones.cc" is pretty large and I think more unit tests will come here.

rongpengzhang replied:
Maybe I can just block the headers that I don't use. Then other developers can just unblock them if they need them. It will be less time-consuming for them.

mbadams5 replied:
@rongpengzhang I understand it is a slight time savings, but really you should strive to only include what you need to include. 

Personally, I think everyone should be explicit about what they need and don't need for unit testing. The unit tests should test very specific functionality and throughly cover all possible valid AND invalid code paths. This is how we build up a large test suite and once that happens we can have much greater confidence when refactoring code to avoid regressions.

Commenting out the includes you do not need for your unit tests works, but I would still push my opinion to remove them entirely.

rongpengzhang replied:
OK. It is fine with me to remove them to be consistent with the unit test custom.

mbadams5 commented on src/EnergyPlus/SimAirServingZones.cc:
```diff
@@ -6317,6 +6361,161 @@ namespace SimAirServingZones {
+		} else if ( ( PrimaryAirSystem( IndexAirLoop ).NumOAHeatCoils > 0 ) || ( PrimaryAirSystem( IndexAirLoop ).NumOAHXs ) ) {
+		//Case: No central heating coils, but preheating coils or OA heat-exchangers exist
+
+			OutAirFrac = FinalSysSizing( IndexAirLoop ).DesOutAirVolFlow / FinalSysSizing( IndexAirLoop ).DesHeatVolFlow;
```
@rongpengzhang Can `FinalSysSizing( IndexAirLoop ).DesHeatVolFlow` ever be 0? If so, there needs to be divide by 0 protection here. It could be as simple as `assert( FinalSysSizing( IndexAirLoop ).DesHeatVolFlow != 0 );` to catch this in debug builds or have any if statement to protect against divide by 0.

rongpengzhang replied:
@mbadams5 Thanks for the careful review. I don't think it can be 0 if it can come here, but I agree it is always good to add a protection to be safer.

mbadams5 commented on src/EnergyPlus/SimAirServingZones.cc:
```diff
@@ -6317,6 +6361,161 @@ namespace SimAirServingZones {
+		} else if ( ( PrimaryAirSystem( IndexAirLoop ).NumOAHeatCoils > 0 ) || ( PrimaryAirSystem( IndexAirLoop ).NumOAHXs ) ) {
+		//Case: No central heating coils, but preheating coils or OA heat-exchangers exist
+
+			OutAirFrac = FinalSysSizing( IndexAirLoop ).DesOutAirVolFlow / FinalSysSizing( IndexAirLoop ).DesHeatVolFlow;
```
@rongpengzhang Same comment as below about divide by 0.

rongpengzhang replied:
I don't think it can be 0 if it can come here, but I agree it is always good to add a protection to be safer.

mbadams5 commented on src/EnergyPlus/SimAirServingZones.cc:
```diff
@@ -1900,6 +1903,29 @@ namespace SimAirServingZones {
+						CompTypeNum = PrimaryAirSystem( AirLoopNum ).Branch( BranchNum ).Comp( CompNum ).CompType_Num;
+
+						if( CompTypeNum == WaterCoil_SimpleHeat || CompTypeNum == Coil_ElectricHeat || CompTypeNum == Coil_GasHeat ) {
+							FoundCentralHeatCoil = true;
```
@rongpengzhang This could maybe be combined with the above (starting line 1874) for loop, but it would require a little refactoring. The quicker change is to at least break from these for loops once `FoundCentralHeatCoil` is found. I don't really like `goto` like the above for loop uses, so i'd probably put `FoundCentralHeatCoil` in the for loop checks themselves. For example,

``` cpp
for( BranchNum = 1; ! FoundCentralHeatCoil && BranchNum <= PrimaryAirSystem( AirLoopNum ).NumBranches; ++BranchNum ) {
  for( CompNum = 1; ! FoundCentralHeatCoil && CompNum <= PrimaryAirSystem( AirLoopNum ).Branch( BranchNum ).TotalComponents; ++CompNum ) {
    CompTypeNum = PrimaryAirSystem( AirLoopNum ).Branch( BranchNum ).Comp( CompNum ).CompType_Num;
    if( CompTypeNum == WaterCoil_SimpleHeat || CompTypeNum == Coil_ElectricHeat || CompTypeNum == Coil_GasHeat ) {
      FoundCentralHeatCoil = true;
    }
  } // end of component loop
} // end of Branch loop
```

This code should work to break early but definitely test it.

rongpengzhang replied:
@mbadams5 I like your latter suggestion! It is a good solution to balance the code efficiency and readability.

mbadams5 commented on src/EnergyPlus/MixedAir.cc:
```diff
@@ -5436,6 +5436,76 @@ namespace MixedAir {
+		for( CompNum = 1; CompNum <= OutsideAirSys( OASysNumber ).NumComponents; ++CompNum ) {
+			CompType = OutsideAirSys( OASysNumber ).ComponentType( CompNum );
+			CompName = OutsideAirSys( OASysNumber ).ComponentName( CompNum );
+			SimOAComponent( CompType, CompName, OutsideAirSys( OASysNumber ).ComponentType_Num( CompNum ), FirstHVACIteration, OutsideAirSys( OASysNumber ).ComponentIndex( CompNum ), AirLoopNum, Sim, OASysNumber, OAHeatingCoil, OACoolingCoil, OAHX );
```
@rongpengzhang Why do you need to call SimOAComponent to set OAHX to then increment NumHX counter? This seems, on the surface, to be an inefficient way to find the number of OA HX. Looking at the `SimOAComponent` code, it seems like the `Sim` boolean makes it so the function basically doesn't do anything, which I think is an odd input parameter for a function called `SimOAComponent`.  This may be the best way to find the number of OA HXs, but it is not intuitive to me. For a seemingly, purpose build function to `GetOASysNumHXs` I think I would take that functionality out of `SimOAComponent` and put it in here.

``` cpp
for( CompNum = 1; CompNum <= OutsideAirSys( OASysNumber ).NumComponents; ++CompNum ) {
  int componentTypeNum = OutsideAirSys( OASysNumber ).ComponentType_Num( CompNum );
  if ( HeatXchngr == componentTypeNum || Desiccant == componentTypeNum ) {
    +NumHX;
  }
}
```

But that is my opinion, others can weigh in if they wish.

On a side note, @DeadParrot I was unaware that `Optional_bool` is a reference, or at least appears to behave like one from this code. Is this true?

DeadParrot replied:
@mbadams5 `Optional` has reference semantics if passed an lvalue of the right type and value semantics otherwise -- you need that to emulate Fortran's OPTIONAL attribute and you want it to avoid copying large objects.

The "native" C++ analog would be having a default value for a (const or non-const) reference argument.

DeadParrot replied:
Semi-minor, nitpicking note: If we are talking about efficiency anyway here is a minor tweak to the proposed loop that is more efficient:

``` cpp
auto const & componentTypeNum = OutsideAirSys( OASysNumber ).ComponentType_Num;
for( CompNum = 1, CompNum_end = OutsideAirSys( OASysNumber ).NumComponents; CompNum <= CompNum_end; ++CompNum ) {
  int const componentNum = componentType_Num( CompNum );
  if ( HeatXchngr == componentTypeNum || Desiccant == componentTypeNum ) {
    +NumHX;
  }
}
```

rongpengzhang replied:
@mbadams5 Umm, another dilemma situation... This solution is definitely more efficient, but may bring more work for  maintenance. I mean, if a new type of component is added in the future, the developer should remember to modify this function, in addition to _SimOAComponent_ which is a collection of many component types.

rongpengzhang committed:
Address Mark's new comments

rongpengzhang commented:
Hi @mbadams5 I have pushed a new commit to address your suggestions. Thanks for the review!

rongpengzhang committed:
Address Mark's new comments

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 83811630_ReheatCoilSizingImprove

rongpengzhang commented:
@Myoldmopar I think this is ready.

jasondegraw commented:
@rongpengzhang One tiny thing: your new unit test doesn't have the license text. Could you add that to the very beginning of the file?

rongpengzhang committed:
Add license text to the new unit test file

rongpengzhang commented:
@jasondegraw Thanks for the kind reminder. It is added.

Myoldmopar commented:
@mbadams5 Do these changes satisfy your notes above?  I could nitpick a bit about how there are some harmless but unnecessary intermediate variables, but I don't feel like holding this up for that type of thing.

mbadams5 commented:
@Myoldmopar Yes he addressed my comments.

mjwitte commented on doc/src/docs/InputOutputReference/01b-InputOutputReference.md:
```diff
@@ -7641,11 +7641,11 @@ Entering *Yes* means the system will be sized for heating using 100% outdoor air
 #### Field: Central Cooling Design Supply Air Humidity Ratio
 
-The design humidity ratio in kilograms of water per kilogram of dry air at the exit of the central cooling coil. (kgWater/kgDryAir) The default is .008.
+The design humidity ratio in kilograms of water per kilogram of dry air at the exit of the central cooling coil. This value is also used for the sizing of zone equipment (e.g., reheat coil) for the system embedded with central heating coils, but it is not used if there is no central heating coil in the system. The default is .008 (kgWater/kgDryAir).
```
@rongpengzhang Sorry to be post-merge with this comment.  Was this sentenced intended to be added to the next field (Central Heating Design Supply Air Humidity Ratio) instead of here?

rongpengzhang replied:
@mjwitte Yes, you are right! Thanks for pointing this out.
